### PR TITLE
H350: Channel-isolated decoder w/ FiLM axis conditioning (WSS_z attack)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The shared `Linear(d_model → 4)` output decoder is the structural bottleneck on WSS_z. All 4 surface channels (cp, τ_x, τ_y, τ_z) read from one linear projection of the same d_model-dim token. τ_z has uniquely streamwise-anisotropic structure (separation/reattachment dominated; on DrivAer it has the largest |τ| variance and the heaviest residual tails — see H346 closure) but must share weight columns with cp/τ_x/τ_y, which have very different spatial statistics. The loss-weight axis is now triply closed (H339 uniform reweight, H341 wz-only, H346 per-vertex focal — all failed without changing this representational competition). Three independent reweighting strategies all bouncing off ~8.62% test_WSS_z is signal: the bottleneck is NOT loss balance, it's that the model has no per-channel decoder capacity.

**Fix:** Replace the shared decoder with **4 axis-conditioned MLP heads** modulated via **FiLM** (Feature-wise Linear Modulation, Perez et al. 2018, arxiv 1709.07871). One learned axis embedding per channel conditions a small MLP head, giving each output its own representation pathway while keeping the encoder shared. Recent geometric ML work (GeoTransolver, Adams et al. Dec 2025, arxiv 2512.20399) shows FiLM-conditioned outputs help anisotropic vector targets in geometric problems.

**Why this should help WSS_z specifically:**
1. τ_z gets dedicated MLP capacity not shared with cp/τ_x/τ_y
2. FiLM is parameter-light: 4 small embeddings + a tiny projection — well under +1% params, so no slope-flattening risk
3. Inductive bias is *output-side*, orthogonal to the encoder/decoder lottery so far
4. Compositional with H336 TTA recipe: if EP15 cosine-tail wins on val, the existing K=5+ν=4+8-res+mirror+cal recipe should compose unchanged

### Implementation (target/`model.py` or wherever the output head lives)

Locate the final output projection in the model (the layer mapping `d_model → 5` for the combined [cp,τ_x,τ_y,τ_z,VP] vector, or `d_model → 4` for surface only — read the current code to verify). Replace it with axis-conditioned FiLM heads:

```python
# In model __init__:
self.axis_embed = nn.Embedding(4, d_model)         # 4 surface channels
self.film_proj = nn.Linear(d_model, 2 * d_model)   # produces (gamma, beta)
self.head_mlp = nn.Sequential(                     # shared MLP, conditioned via FiLM
    nn.Linear(d_model, d_model // 2),
    nn.GELU(),
    nn.Linear(d_model // 2, 1),
)
# Keep the existing volume head (VP) untouched — only refactor surface.

# In forward (after the per-token d_model representation h is computed):
# h: (B, N, d_model) — surface tokens
B, N, D = h.shape
axis_ids = torch.arange(4, device=h.device)            # (4,)
axis_e = self.axis_embed(axis_ids)                     # (4, d_model)
gb = self.film_proj(axis_e)                            # (4, 2*d_model)
gamma, beta = gb.chunk(2, dim=-1)                      # each (4, d_model)
# Broadcast: (B, N, 1, D) * (1, 1, 4, D) + (1, 1, 4, D) -> (B, N, 4, D)
h_cond = h.unsqueeze(2) * (1.0 + gamma.view(1, 1, 4, D)) + beta.view(1, 1, 4, D)
surface_out = self.head_mlp(h_cond).squeeze(-1)        # (B, N, 4)
# VP head unchanged, concat as before to make (B, N, 5).
```

Verify the FiLM gamma is initialized so the head starts near-identity:
- `nn.init.zeros_(self.film_proj.weight)` and `nn.init.zeros_(self.film_proj.bias)` → gamma=0, beta=0 at init → `h_cond = h` → equivalent to original shared head for first few steps.

### Training (diagnostic gate first — 6h frozen-backbone)

We do NOT want to commit to a 24h-cosine-tail train on something this architectural without a fast falsification check.

**Phase A (diagnostic, ~6h on 8×A100 DDP):** Load H336 source checkpoint EP13 (W&B artifact `yw2a5dyl:epoch-13`), freeze the encoder, train only the new FiLM heads + the volume head from scratch for 3 epochs (warm-restart cosine, lr=3e-4, T_max=3). This tests whether the FiLM decoder can match or beat the original shared decoder on top of a fixed backbone.

```bash
torchrun --standalone --nproc-per-node=8 train.py \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 \
  --epochs-already-done 13 --epochs 16 --lr-cosine-t-max 16 \
  --freeze-backbone \
  --film-decoder \
  --batch-size 8 --num-workers 4 \
  --wandb_group h350-askeladd-filmdec-frozen \
  --wandb-name askeladd/h350-phase-a-frozen-filmdec
```

**Gate to Phase B:** After Phase A run completes, post `val_abupt`, `val_WSS_z`, and the 5 test channels in a PR comment. Proceed to Phase B (full cosine-tail finetune) ONLY if `val_WSS_z` improves by ≥5bp vs H336 (i.e. val_WSS_z < expected H336 val_WSS_z value — query W&B run `348i3z1v` for exact numbers). Otherwise close with finding `axis-cond-decoder-frozen-null`.

**Phase B (full finetune, if gate passes, ~20-24h on 8×A100 DDP):**
```bash
torchrun --standalone --nproc-per-node=8 train.py \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 \
  --epochs-already-done 13 --epochs 16 --lr-cosine-t-max 16 \
  --film-decoder \
  --batch-size 8 --num-workers 4 \
  --wandb_group h350-askeladd-filmdec-full \
  --wandb-name askeladd/h350-phase-b-full-filmdec
```

**Evaluation (Phase B, after training):** Use the H336 TTA recipe exactly so the result is comparable to current SOTA:
```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-<PHASE-B-RUN-ID>/checkpoint_ep15.pt \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 5 --antithetic-noise \
  --weight-noise-dist student_t --weight-noise-df 4 \
  --test-time-calibration \
  --batch-size 2 --num-workers 4
```

### Constraints (must preserve)

- DDP with 8 GPUs (use `torchrun --nproc-per-node=8`)
- z-axis `tau_z` LOCKED at the trained value (1.67) — do NOT modify
- Do NOT touch `data/loader.py`, `data/preload.py`, `data/split_manifest.json`
- Param overhead MUST be < +1% (FiLM heads should be well under this — verify with parameter count print at training start)
- Report exact param-count delta vs H336 in PR comment

## Baseline

Current SOTA: **H336 K=5 + Student-t ν=4 + 8-res + mirror + cal** (PR #1520, merged to `tay` 2026-06-01)

- val_abupt_calibrated: **5.8978%**
- test_abupt_calibrated: **5.7379%**
- test_VP: 3.3735%
- test_SP: 3.6133%
- test_WSS: 6.6382% (target: < 5.85% — Morgan Issue #1056)
- test_WSS_x: 5.9014%
- test_WSS_y: 7.1841%
- test_WSS_z: **8.6175% ← primary target for this hypothesis**

**Merge gate:** `val_abupt_calibrated < 5.8978%` AND `test_abupt_calibrated < 5.7379%`

**Per-channel calibration (apply same coefficients at eval):**
- surface[cp]: α=+0.99476, β=−0.00236
- surface[τ_x]: α=+0.99443, β=−0.00721
- surface[τ_y]: α=+0.99425, β=−0.00023
- surface[τ_z]: α=+0.99172, β=−0.00018
- volume[VP]: α=+0.99963, β=−0.86162

**Source checkpoint:** H185 EP15 EMA = `outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt` (EP13 base = W&B artifact `yw2a5dyl:epoch-13`)

**SOTA W&B run:** `348i3z1v` (nezuko/h336-K5-studentt-nu4-8res-cal)

**Falsification trigger:** if Phase A (frozen backbone) does NOT improve `val_WSS_z` by ≥5bp vs H336, close with finding `axis-cond-decoder-frozen-null` and we'll know the WSS_z floor is encoder-bottlenecked, not decoder-bottlenecked.

## Context: H346 closure motivates this attack

Just closed PR #1537 (H346 per-vertex focal WSS_z reweighting). Combined with H339 (uniform reweight) and H341 (wz-only reweight), the **loss-reweight axis is triply closed**. Banked finding: `wss-z-loss-weight-axis-closed`. WSS_z at 8.62% test_cal is NOT a gradient-budget problem — it's a representational/capacity problem. This hypothesis attacks the architecture, not the loss.

Companion in-flight architecture work this round:
- fern H348: surface curvature input features (k1, k2, H, K) — input-side
- frieren H349: arcsinh target transformation — output-transform side
- askeladd H350 (this PR): axis-conditioned output decoder — output-decoder side

Each addresses a different mechanism; collectively they triangulate where the WSS_z floor sits.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd:

Phase A launched on 8×A100 DDP. W&B rank-0 run: `fq2m1g8c` (group `h350-askeladd-filmdec-frozen`).

## Implementation notes

**Decoder spec deviations from PR body (all logged in code comments):**
- `axis_embed` dim **=32** (not d_model=512). At H336 scale this lands the FiLM heads at **−99,203 net params (−0.57% of base)**; the PR's dim=d_model variant would have been **+394,237 params (+2.26%)**, violating the ≤+1% constraint. Channel identity only needs 4 discrete embeddings, so a low-dim signature loses no expressivity (effective rank of the FiLM modulation is bounded by 4 distinct axis embeddings, not by axis_dim).
- Forward uses a per-channel loop, not the `(B, N, 4, D)` broadcast — at H336 (N≈65k, D=512, B=8, C=4) the broadcast tensor is ~4 GB per device and risks OOM; the loop reuses a `(B, N, D)` buffer at 4× sequential cost. Verified equivalent.
- `film_proj.weight` and `film_proj.bias` zero-init (γ=0, β=0 at step 0) — `head_mlp` standard trunc_normal init.
- Volume head retained unchanged (loaded from EP13 checkpoint).

**Schedule deviation from PR bash (description aligned):**
- PR description says *"warm-restart cosine, lr=3e-4, T_max=3"*; PR bash had `--epochs-already-done 13 --epochs 16 --lr-cosine-t-max 16` which would put the LR at ~2.6e-5 at start of the new training — too low for randomly-initialised head_mlp. Going with **fresh 3-epoch cosine at lr=3e-4** per the description (`--epochs 3 --lr-cosine-t-max 3 --epochs-already-done 0 --lr-warmup-epochs 0`). Other hyperparams match H336 source training run `0gjfv45i` exactly: weight_decay=5e-4, grad_clip=0.5, EMA(0.999, start=50), mirror_augmentation, save_every_epoch, tau_y=1.3, tau_z=1.67.

**Freeze logic:** all 16,984,660 backbone+encoder params frozen; 329,858 trainable (1.91%): `axis_embed`, `film_proj`, `head_mlp` (new) + `volume_out` (preserved, unfrozen per PR's "FiLM heads + the volume head" wording). Optimizer only sees trainable params; DDP enables `find_unused_parameters=True` for the frozen graph.

**Checkpoint load:** EP13 EMA loaded with `missing=7` (new FiLM keys: `axis_embed.weight`, `film_proj.{weight,bias}`, `head_mlp.{0,2}.{weight,bias}`), `unexpected=4` (old `surface_out.{0,2}.{weight,bias}` discarded). Verified `film_proj` zero-init at step 0.

## Gate

H336 source training `0gjfv45i` raw single-checkpoint (no TTA) val_WSS_z = **9.1839%** (test_WSS_z = 8.7551%). Gate: Phase A `val_primary/wall_shear_z_rel_l2_pct` must improve by ≥5bp → target **< 9.1339%**.

Will update with results once Phase A finishes (~3 epochs frozen-backbone, ~3-6h estimated).

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1542 because ADVISOR (closing 2026-06-01 16:55Z): Phase A frozen-backbone diagnostic FAILS the student-committed gate catastrophically.

**Phase A actual (rank0 `fq2m1g8c`, EP3 EMA):**

| Metric | Phase A | H336 baseline | Δ |
|---|---:|---:|---:|
| val_abupt | **8.9393** | 5.9686 raw | **+297bp catastrophic** |
| val_wss_z | **12.3415** | 9.1839 raw | **+316bp catastrophic** |
| val_wss_y | 12.2302 | 7.18 | +505bp |
| val_wss_x | 9.4354 | 5.93 | +351bp |
| val_SP | 7.1494 | 3.94 raw | +321bp |
| val_VP | 3.5398 | 3.40 raw | +14bp (volume head preserved) |

**Student's pre-committed gate**: `val_primary/wall_shear_z_rel_l2_pct < 9.1339%` (5bp improvement over H336 raw). Actual: **12.34% = +316bp worse**. Gate HARD FAIL.

## Mechanism finding: `filmdec-random-init-too-slow-for-frozen-backbone`

The Phase A failure isolates to **head_mlp random initialization**, not to the FiLM-decoder architectural direction itself. The student explicitly discarded EP13's `surface_out.{0,2}.{weight,bias}` (per their setup comment: "unexpected=4 ... discarded") and re-initialized `head_mlp` from scratch with trunc_normal init.

With backbone frozen and 3 epochs at lr=3e-4 fresh cosine, the randomly initialized head_mlp could not recover EP13's surface prediction behavior. The volume head was preserved (only +14bp val_VP regression), confirming the diagnosis: any new randomly-init surface head will lose ~300bp vs the trained EP13 baseline at this schedule.

**This does NOT falsify FiLM-decoder direction.** It falsifies the specific "random-init head_mlp + 3-epoch frozen-backbone" approach. Two paths forward:

1. **Warm-start head_mlp from EP13 `surface_out`** (architectures match: both are 2-layer MLPs at indices 0, 2). With γ=β=0 FiLM init, step-0 prediction is EXACTLY EP13 — so Phase A' would START at val_wss_z ≈ 9.18% and only need to NOT regress. This is the correct diagnostic gate.
2. **Skip Phase A and run Phase B full 16-epoch cosine-tail** — but commits ~24h GPU time on a hypothesis whose head init has been shown to be slow.

Path 1 is strictly cheaper and answers the question (does FiLM modulation help once heads are properly initialized?). Assigning H354 as the corrected Phase A' to askeladd next.

## Banked findings
- `filmdec-phasea-random-init-too-slow-frozen-backbone` — randomly initializing surface decoder heads (head_mlp 329k params) cannot recover EP13's val_wss_z in 3 epochs even with backbone frozen. Surface heads need warm-start, not from-scratch.
- `volume-head-preserved-when-untouched` — preserving the EP13 volume head verbatim kept val_VP regression within 14bp, confirming the regression is heads-only (not backbone or any cascading effect).

Closing PR #1542. H354 corrected-FiLM assigned next..
